cfgov/regulations3k/scripts/ecfr_importer.py

```diff
@@ -0,0 +1,282 @@
+            contents=section_content
+        )
+        _section.save()
+
+
+def parse_appendices(appendices_list, part):
+    """
+    Parse the list of appendices, tease out interps and send them along,
+    and then process the remaining appendices.
+    """
+    PAYLOAD.appendices = {}
+
+
+def parse_interps(interp_list, part):
+    """
+    Take the remaining list of interpretations, break them down into
+    individual section interpretations, and process them by subpart.
+    """
+    PAYLOAD.interpretations = {}
+
+
+def ecfr_to_regdown(part_number, file_path=None):
+    """
+    Extract a regulation Part from eCFR XML, and create regdown content.
+
+    The default XML source is the latest regulation posting at www.gpo.gov,
+    which gets updated every few days.
+
+    If `file_path` is specified, a local XML file is parsed instead.
+
+    DIV1 is a title (as in Title 12)
+    DIV3 is a chapter (not used here)
+    DIV5 is a part
+    DIV6 is a subpart
+    DIV8 is a section
+    DIV9 is an appendix
+    DIV9 element whose HEAD starts with 'Supplement I' is an interpretation
+    """
+    starter = datetime.datetime.now()
+    if file_path:
+        try:
+            with open(file_path, 'r') as f:
+                markup = f.read()
+        except IOError:
+            logger.info("Could not open local file {}".format(file_path))
+            return
+    else:
+        ecfr_request = requests.get(LATEST_ECFR)
+        if ecfr_request.reason != 'OK':
+            logger.info(
+                "ECFR request failed with code {} and reason {}".format(
+                    ecfr_request.status_code, ecfr_request.reason))
+            return
+        ecfr_request.encoding = 'utf-8'
+        markup = ecfr_request.text
+    soup = bS(markup, "lxml-xml")
+    parts = soup.find_all('DIV5')
+    try:
+        part_soup = [div for div in parts if div['N'] == part_number][0]
+    except IndexError:
+        logger.info(
+            "Regulation could not be found for part {}".format(part_number))
+        return
+    part = parse_part(part_soup, part_number)
+    parse_version(part_soup, part)
+    subpart_list = part_soup.find_all('DIV6')
+    parse_subparts(part_soup, subpart_list, part)
+    appendices_list = part_soup.find_all('DIV9')
+    interps_list = [div for div in appendices_list
+                    if div.find('HEAD').text.startswith('Supplement I')]
+    parse_appendices(appendices_list, part)
+    parse_interps(interps_list, part)
+
+    msg = (
+        "Draft version of Part {} created.\n"
+        "Parsing took {}".format(
+            part_number, (datetime.datetime.now() - starter))
+    )
+    return msg
+
+
+def run(*args):
+    if len(args) not in [1, 2]:
+        logger.info(
+            "Usage: ./cfgov/manage.py runscript "
+            "ecfr_importer --script-args "
+            "[PART NUMBER] [OPTIONAL XML FILE PATH]")
+        sys.exit(1)
+    elif len(args) == 1:
```

Comment:
I wonder, since we have to download all of Chapter X anyway, could we make PART NUMBER optional, and if it's not given, import all parts?

Reply:
I do fear admin pollution, as an unwanted full download would require deleting versions for every reg, and stakeholders have expressed a desire to limit the authority to delete.  
Maybe get some feedback from stakeholders/editors on who has authority to download and who has authority to delete?

Reply:
That's a good point... are we going to hook this up to a button in the admin?

Reply:
I would vote for a button and mini API, but I think we need some feedback.

Reply:
![button](https://media.giphy.com/media/3osxY9xCBxlndni2XK/giphy.gif)